### ISSUE TITLE
[key-spacing] Allow for "alignment" spacing configuration

### ISSUE THREAD

**Issue by Kyle-Mendes** [eslint/eslint#5613]
**What version are you using?**
v2.4.0

**What did you do?**
Try to configure the `key-spacing` rule to my desired configurations.

```json
"key-spacing": [1, {
  "singleLine": {
    "beforeColon": false,
    "afterColon": true
  },
  "multiLine": {
    "align": "colon",
    "mode": "minimum"
  }
}],
```

**What happened?**
See below.

**What did you expect to happen?**
See below. [**new feature / enhancement**]

----------------

I'm trying to configure the `key-spacing` spacing rule to work as follows:

* **Single-line definitions**: No space before colon, space after colon .
* **Multiple-line definitions**:
  * IF the multiple-line definition would align another definition: 
    * space before & after colon
    * align on colon.
  * Otherwise: No space before colon, space after colon.

Valid Examples:

```javascript
var options = {
    callback: function(cb) { // no space before colon, 1 after
        cb();
    }
};

var obj = { // align on colon, space before and after colon. 
    a   : 1,
    bc  : 2,
    def : 3, // The space-after colon is 1 space after the longest key
    g   : 4
};
```

As far as I can tell, there is no way to enforce this styling.  I can either get eslint to warn that the callback example is missing a space before the colon, or that the object example has 1 extra space after `def`.  I can't get a configuration that throws no warnings.

---------------

I think there should be an additional `alignment` configuration for `key-spacing` to further allow users to define fined-grained control over the alignment.

Example:
```json
"key-spacing": [2, {
    "singleLine": {
        "beforeColon": false,
        "afterColon": true
    },
    "multiLine": {
        "beforeColon": false,
        "afterColon": true,
        "align": "colon"
    },
    "align": {
        "beforeColon": true,
        "afterColon": true
    }
}]
```

I believe a configuration like this would accomplish the exact type of styling that I'm looking for.

**Comment by btmills** [comment]
@Kyle-Mendes to clarify, how would you style the following?

```js
var obj = {
    callback: function(cb) {
        cb();
    },
    a: 1,
    bc: 2
};
```

**Comment by Kyle-Mendes** [comment]
Great question @btmills ! 

In this instance, the way I would break it up would to treat the `{}` of the callback function as a delimiter breaking up the object into two "groups" of key values.  If I'm not mistaken, eslint will already do that in regards to the `align` config.

The result would be:

```js
var obj = {
    callback: function(cb) {
        cb();
    },
    a  : 1,
    bc : 2
};
```

Complex: 
```js
var obj = {
    _origin  : 0,
    _x       : 1,
    callback : function(cb) {
        cb();
    },
    a  : 1,
    bc : 2
    d  : function() {
        //...
    },
    o: {
        internal : true,
        int      : 1
    }
};
```

So, each "group" of values would follow the rules for single line, multi-line, or aligned spacing.

**Comment by nzakas** [comment]
@btmills can you comment on this?

My initial feeling is that this would be pretty complicated, but I don't know the code well enough.

**Comment by ilyavolodin** [comment]
We had some trouble with alignment in the past. Mostly around figuring out how to group things for alignment (I think it was in `one-var` or at least some rule that dealt with styling around variable declarations). I think it would be a reasonable addition, but I'm afraid there might be too many edge-cases.

**Comment by btmills** [comment]
@Kyle-Mendes sorry this got lost, that's my fault.

As to feasibility, I think we got lucky on this one and it might be among the easier changes to `key-spacing`. We're already [separating properties into groups](https://github.com/eslint/eslint/blob/46d061059dfb7c07caca478adc552658ce70e09a/lib/rules/key-spacing.js#L392), so In Theory&trade;, when [`verifyGroupAlignment` determines `beforeColon` and `afterColon` settings](https://github.com/eslint/eslint/blob/46d061059dfb7c07caca478adc552658ce70e09a/lib/rules/key-spacing.js#L361-L362), it could check whether `properties.length > 1`.

I agree with @ilyavolodin that this is something that's reasonable to add. @nzakas are you willing to accept it once we spec it out fully? @Kyle-Mendes your examples show pretty clearly what's expected, but I fear having an `align` key at the same level as `singleLine` and `multiLine` could be confusing since it only applies to multi-line literals. I'd like to move the option inside `multiLine` somewhere to make that clear. Any ideas? `{ "multiLine": { "align": { "part": "colon|value", "beforeColon": false, "afterColon": true } } }` seems like it adds too many levels of nesting.

@Kyle-Mendes would you be willing to implement this if it's accepted? I'm happy to provide guidance as necessary.

**Comment by pedrottimark** [comment]
Thinking about describing the option schema, something that I cannot tell for sure from https://github.com/eslint/eslint/issues/5613#issue-141943387 is whether the heart of the request is
* multiLine options apply to “groups” with more than one property
* singleLine options apply not only to single-line object literal but also to groups with only one property

If it is not more complicated than that one possibility that comes to mind is a boolean property in the singleLine object which means that config takes precedence over the multiLine config for properties in singleton groups.

**Comment by btmills** [comment]
@pedrottimark I'm so glad you read these. My comprehension was clouded by being too familiar with the implementation. Going back now and re-reading, this is what I'm getting:

- Currently, `singleLine` and `multiLine` apply based on whether the properties in the *entire object* spans more than one line.
- Assuming @pedrottimark is correct, this issue wants an option to change the behavior such that `singleLine` and `multiLine` would apply based on whether the properties in the *individual property group* span more than one line - singleton property groups would therefore only be subject to the `singleLine` options.

@Kyle-Mendes is that right? The underlying implementation would be the same, but the option schema and docs would both be much simpler.

**Comment by Kyle-Mendes** [comment]
I took a stab at implementing this here: https://github.com/Kyle-Mendes/eslint/commit/a0d23e766ea84bab73c28196b9adfb4d0a65dfad. 

@pedrottimark Was close, but simply using the single-line spacing isn't exactly what I was looking for.  The expanded example I gave wouldn't be possible with that, I don't believe. 

What I did was add an `alignment` property which allows for additional configuration for the rule that only takes effect when a group is going to be aligned.  This allows the alignment rule to be completely independent of `singleLine` and `multiLine`. 

I'm planning on adding to these tests further to make sure I didn't miss anything, but it doesn't appear that my changes should be breaking anything in the API, since all previous tests are still passing.

**Comment by pedrottimark** [comment]
Good timing that the doc for this rule is in the next batch to be updated, so we can think about the description along with the implementation and the examples along with the tests.

@scriptdaemon If you have already started or will soon start to think about key-spacing.md this is an issue to watch and your thoughts are welcome. If you prefer not (and that’s okay) let me know.

**Comment by scriptdaemon** [comment]
@pedrottimark I have started, and I was thinking I could have the next batch ready for a PR tomorrow.

However, do we want to wait for this? I'm wondering if I can get the next batch in a ready state before this issue is resolved, I can update the doc for this rule as it currently works. Then, in the PR for this issue, we can discuss the changes corresponding to the updates and worry about it at that time. What do you think?

**Comment by pedrottimark** [comment]
@scriptdaemon Sorry I was unclear to sound like asking you to wait. Am glad if we merge improvements to the rule doc with current options, to see more clearly how the new option will fit.

**Comment by pedrottimark** [comment]
@Kyle-Mendes @btmills Here is an odd thought. Especially if this rule becomes **fixable** in the future, could conditional behavior related to single or multiple property alignment group (which I am still trying to understand correctly;) cause a spacing change to an adjacent property when you add a new property?

That is, could it have the side-effect of causing a problem analogous to what comma-dangle solves? But I guess that is a general effect of some alignment styles?

Taking the examples in https://github.com/eslint/eslint/issues/5613#issuecomment-198534580

```
var obj = {
    callback: function(cb) {
        cb();
    },
    a  : 1,
    bc : 2
};
```

Adding `_x` causes (by fixing or reporting) change to spacing of `callback` as `bc` would cause to `a`.

```
var obj = {
    _x       : 1,
    callback : function(cb) {
        cb();
    },
    a  : 1,
    bc : 2
};
```

**Comment by Kyle-Mendes** [comment]
@pedrottimark Yep, that's correct.  The additions of properties can force a code change on other lines if your `alignment` configuration didn't match either the `multiLine` or `singleLine`.  In your example above, if `multiLine` was set to `beforeColon: true, afterColon: true` then there wouldn't be a line change. 

I believe that it is possible to trigger the spacing change to an adjacent property with the current configuration as well, if you have different `singleLine` than your `multiLine` and have `align: "colon"`

**Comment by btmills** [comment]
> @Kyle-Mendes I believe that it is possible to trigger the spacing change to an adjacent property with the current configuration as well

Yep, and it doesn't even require separate `singleLine` and `multiLine` config:

```js
/* eslint key-spacing: [2, { "align": "colon" }] */
var obj = {
    a: b
}
```

Add a longer key, and you have to adjust every property in the group:

```js
var obj = {
/* eslint key-spacing: [2, { "align": "colon" }] */
    a  : b,
    baz: bat
};
```

> @pedrottimark But I guess that is a general effect of some alignment styles?

Precisely. I don't think this option would introduce anything there that we wouldn't already have to handle.

**Comment by btmills** [comment]
@Kyle-Mendes your implementation in https://github.com/Kyle-Mendes/eslint/commit/a0d23e766ea84bab73c28196b9adfb4d0a65dfad demonstrates that this is practical to implement. I'd still like to tweak your proposed options a bit to make it clear that the cases where the new option applies are a subset of those where alignment might apply. Brainstorming, what about this?

```json
{
    "align": "colon",
    "individuals": {
        "beforeColon": false,
        "afterColon": true
    },
    "groups": {
        "beforeColon": true,
        "afterColon": true
    }
}
```

In the `singleLine`/`multiLine` case, you'd have this, which would allow you to configure single-line objects and individual properties on multi-line objects separately (am I reading correctly that you want to be able to do that?).

```js
{
    "singleLine": {
        "beforeColon": false,
        "afterColon": true
    },
    "multiLine": {
        "align": "colon",
        "individuals": {
            "beforeColon": false,
            "afterColon": true
        },
        "groups": {
            "beforeColon": true,
            "afterColon": true
        }
    }
}
```

**Comment by Kyle-Mendes** [comment]
@btmills 
> I'd still like to tweak your proposed options a bit to make it clear that the cases where the new option applies are a subset of those where alignment might apply.

I think that's a great idea.  And I really like the idea of `individuals` and `groups`.  My only worry is that it's not immediately semantically clear, and also would introduce a breaking API change for current users. 

What if we did something like this?

```json
{
	"singleLine": {
		"beforeColon": false,
		"afterColon": true
	},
	"multiLine": {
		"align": {
			"on": "colon", // Not sure what to call this property.  
			"beforeColon": true,
			"afterColon": true
		},
		"beforeColon": false,
		"afterColon": true
	}
}
```

Then, the schema would only need to change to make `align` accept a string or an object.  If it's an object, we use the logic that I added in the PR.  This wouldn't break anyone's upgrade to new versions of eslint, and I think that it still imparts a semantic understanding of what the configurations are doing at-a-glance.

**Comment by zxqfox** [comment]
I think we should this issue to JSCS Milestone: http://jscs.info/rule/requireAlignedObjectValues ;-)

**Comment by Kyle-Mendes** [comment]
@zxqfox Does this allow for the flexible configuration that we have here though?  That matches my personal use case, but I think there's something to be said about the ability to set the specific alignment options like we have above.

**Comment by zxqfox** [comment]
Not so flexible as you already invented but there are two specific options for grouping records dividing by newlines or functions:
```js
var a = {
  x   : 1,
  abc : 2,
  step: 3,

  another: 1,
  group  : 2,

  makeMyDayBrighter: function() {
  },
  anotherMethod: function() {
  }
};
```

Would be nice if you will bear in mind this.
Also you can take a look at specs: https://github.com/jscs-dev/node-jscs/blob/master/test/specs/rules/require-aligned-object-values.js.

Oh, there are also computed property names and methods:
```js
var k = 'yo';
var x = {
  [k]: 1,
  k  : 2,
  m() {}
}
```

**Comment by btmills** [comment]
@zxqfox it looks like the JSCS equivalent is sufficiently different from this proposal to deserve its own issue. Can you create one outlining what you think compatibility should look like? We can add that to the milestone.

**Comment by btmills** [comment]
@Kyle-Mendes no matter which way we do this, it won't need to be a breaking change. I'm fine with `on` as the key name for align, "align on colons" or "align on values" makes sense to me.

Your proposal is this:

```json
{
    "singleLine": {
        "beforeColon": false,
        "afterColon": true
    },
    "multiLine": {
        "align": {
            "on": "colon",  
            "beforeColon": true,
            "afterColon": true
        },
        "beforeColon": false,
        "afterColon": true
    }
}
```

Would we also support the object `align` value in the combined config like this?

```json
{
    "beforeColon": false,
    "afterColon": true,
    "align": {
        "on": "colon",  
        "beforeColon": true,
        "afterColon": true
    }
}
```

**Comment by Kyle-Mendes** [comment]
I think we should include the `align` object as well.  It shouldn't be too difficult to support.  I'd be happy to take a stab at implementing this new syntax and fixing the PR if you'd like.

**Comment by btmills** [comment]
I'm happy with the new options as shown in https://github.com/eslint/eslint/issues/5613#issuecomment-212526880. We would of course maintain backwards compatibility with string values of "colon" or "value" for "align" as well. Anyone else have any thoughts?

**Comment by ilyavolodin** [comment]
Sounds reasonable to me. Although the schema is going to look very complex for this rule, but I see no other good option.

**Comment by Kyle-Mendes** [comment]
@btmills @ilyavolodin how do we want to handle `mode` for the new `align` configuration?  Previously, it was a property of just `multiLine`.  Should it now also be a parameter for the top level `align`?

**Comment by btmills** [comment]
> Should it now also be a parameter for the top level align?

Good catch. I think it should.

**Comment by Kyle-Mendes** [comment]
@btmills @ilyavolodin @pedrottimark @zxqfox  (sorry for @'ing everyone, not sure who to include)

I found and interesting interaction in the `key-spacing` rule, and I'm not sure if it's intended or not.  

Object declarations like this: 

```js
{
    key: "value"
}
```

Are being classified as a "group" declaration, so they get the rules for `mulitLine` applied when checking spacing.  The following rules will result in `Extra space before value for key ‘key’.` being thrown when checking the object.

```js
singleLine: {
    beforeColon: false,
    afterColon: true
},
multiLine: {
    beforeColon: false,
    afterColon: false
}
```

To me, this would be a `singleLine` declaration, since there is only one property, and it is on a single line.  Whereas the following would be a `multiLine` declaration, since the single property spans multiple lines.

```js
// multLine
{
    f: function() {
        // stuff
    }
}
```

```js
// complex
{
    key: "value",             // group
    things: "are happening",  // group
    f: function() {           // group
        var x = {
            f2: function() {  // multiLine

            }
        }
    },
    singleLine: "declaration" // singleLine
}

```

**Comment by Kyle-Mendes** [comment]
@btmills and I were chatting on gitter and had a decision that we thought would benefit from some outside perspective. Take the following example:

```js
// config
{
    align: {
        on: "colon",
        afterColon: true
    },
    beforeColon: true,
    afterColon: false
}

// code
{
    longName: 1,
    small   : 2,
    f       : function() {
        // function
    },
    xs :3
}
```

Would we expect that the rule throws an error for not having a space before the colon for the alignment options?  There's two options

1) Undeclared `beforeColon` values default to `false`.  Since the `align` object doesn't have a value for `beforeColon`, it defaults to `false`, and thus the above code is free of errors.
2) `align` inherits the `beforeColon` value from the default configuration and thus, the code throws an error for missing spaces `beforeColon`
  - A consequence of this would be that `align` should inherit from `multiLine` as well but potentially **only** when the `align` property is nested under `multiLine`? If not, it would only ever inherit from top-level `before/afterColon` values.

I am personally of the mind that it should not inherit, while @btmills thinks it should!

Either way, I don't think there's an option that will be inherently apparent to all users, so we will need to make sure to make careful note of the decision we come to in the documentation!

**Comment by nzakas** [comment]
Is someone championing this change? It's not yet "accepted", but it looks like there's been a bunch of work on a pull request.

**Comment by btmills** [comment]
:hand: I will.

@Kyle-Mendes initially submitted the PR to show that this is a small change to the implementation. It's also helping us hash out the right configuration, which is still a bit in flux. Hopefully we can get some feedback and come back with a final proposal for acceptance.

**Comment by Kyle-Mendes** [comment]
@btmills Is all that is outstanding the documentation and the above change (inheriting spacing config)?

**Comment by btmills** [comment]
@Kyle-Mendes I believe so.

@eslint/eslint-team what do you think about https://github.com/eslint/eslint/issues/5613#issuecomment-216624115?

**Comment by nzakas** [comment]
I'll be honest, this rule is pretty far above my head, so I'm not sure my opinion is helpful. I'll defer to @btmills opinion.

**Comment by ilyavolodin** [comment]
I'll support this, :+1: Need a few more from @eslint/eslint-team

**Comment by mikesherov** [comment]
:+1:

**Comment by nzakas** [comment]
:+1:
